src/cargo/ops/cargo_install.rs

```diff
@@ -200,31 +200,35 @@ fn install_one(
         )?
     };
 
-    let mut td_opt = None;
-    let mut needs_cleanup = false;
-    let overidden_target_dir = if source_id.is_path() {
-        None
-    } else if let Some(dir) = config.target_dir()? {
-        Some(dir)
-    } else if let Ok(td) = TempFileBuilder::new().prefix("cargo-install").tempdir() {
-        let p = td.path().to_owned();
-        td_opt = Some(td);
-        Some(Filesystem::new(p))
+    let (mut ws, git_package) = if source_id.is_git() {
+        // Don't use ws.current() in order to keep the package source as a git source so that
+        // install tracking uses the correct sourc.
+        (Workspace::new(pkg.manifest_path(), config)?, Some(&pkg))
+    } else if source_id.is_path() {
+        (Workspace::new(pkg.manifest_path(), config)?, None)
     } else {
-        needs_cleanup = true;
-        Some(Filesystem::new(config.cwd().join("target-install")))
-    };
-
-    let mut ws = match overidden_target_dir {
-        Some(dir) => Workspace::ephemeral(pkg, config, Some(dir), false)?,
-        None => {
-            let mut ws = Workspace::new(pkg.manifest_path(), config)?;
-            ws.set_require_optional_deps(false);
-            ws
-        }
+        (Workspace::ephemeral(pkg, config, None, false)?, None)
     };
     ws.set_ignore_lock(config.lock_update_allowed());
-    let pkg = ws.current()?;
+    ws.set_require_optional_deps(false);
```

Comment:
I'd love to know how you feel about this specific change @ehuss. I feel like generally it's a bit more Rust-y, but it does come with the disadvantage that the `set_require_optional_deps(false)` is basically a pointless no-op for the ephemeral workspace since that is already created with that argument.

But generally I feel like this slims down the different cases a bit and makes it a little more obvious maybe that we're looking at three mostly different variations, rather than two variation where one has another sub-variation? Both seem fine, but I think I'd personally prefer this approach.

This of course also has the benefit that the `git_package` is not mutable anymore and only ever created with the value that it's also going to be used with.